### PR TITLE
ENH: expose get_coordinates as a method

### PR DESCRIPTION
**Pull request by martinfleis**
This exposes `get_coordinates` from shapely2/pygeos as a GeoSeries/GeoDataFrame method, returning a pandas.DataFrame. The index handling mirrors the implementation we have in `explode` as both are very similar operations in nature. One is returing parts of original geometries, the other goes one step further and returns coordinates.

I am aware that this is a borderline function we weren't sure we want to expose as a method but given there is a relatively close precedent with `explode` and we can reuse its indexing machinery, it is in the end a nice addition I think. 

I will add tests and documentation once there's an agreement that we want to do this and we have the API set.

Current behaviour:
```py
>>> from shapely.geometry import Point, LineString, Polygon
>>> s = geopandas.GeoSeries(
...     [
...         Point(1, 1),
...         LineString([(1, -1), (1, 0)]),
...         Polygon([(3, -1), (4, 0), (3, 1)]),
...     ]
... )
>>> s
0                              POINT (1.00000 1.00000)
1       LINESTRING (1.00000 -1.00000, 1.00000 0.00000)
2    POLYGON ((3.00000 -1.00000, 4.00000 0.00000, 3...
dtype: geometry
```
```py
>>> s.get_coordinates()
     x    y
0  1.0  1.0
1  1.0 -1.0
1  1.0  0.0
2  3.0 -1.0
2  4.0  0.0
2  3.0  1.0
2  3.0 -1.0
```
```py
>>> s.get_coordinates(ignore_index=True)
     x    y
0  1.0  1.0
1  1.0 -1.0
2  1.0  0.0
3  3.0 -1.0
4  4.0  0.0
5  3.0  1.0
6  3.0 -1.0
```
```py
>>> s.get_coordinates(index_parts=True)
       x    y
0 0  1.0  1.0
1 0  1.0 -1.0
  1  1.0  0.0
2 0  3.0 -1.0
  1  4.0  0.0
  2  3.0  1.0
  3  3.0 -1.0
```

### CODE REVIEW

File: geopandas/base.py [hunk 1]
```diff
@@ -3450,6 +3451,136 @@ def equals(self, other):
...
+            import pygeos
+
+            coords, outer_idx = pygeos.get_coordinates(
+                self.geometry.values.data, include_z=include_z, return_index=True
+            )
+
+        else:
+            raise NotImplementedError(
+                f"shapely >= 2.0 or PyGEOS is required, "
+                f"version {shapely.__version__} is installed"
```
Comment (m-richards) [inline geopandas/base.py:3540]: I know you haven't written tests yet so this probably hasn't shown up, but this block will need to import shapely

**Review by m-richards — COMMENTED** [review]
I think this would be nice to expose on the geopandas level in addition to shapely - especially given we already have the logic to do so worked out thanks to `explode`. I think it would be nice if we could abstract and reuse the same indexing code rather than reimplementing it to ensure the two pieces of logic stay in sync (especially given that it's complicated and there's already been a few bugs found in explode) but not critical. 

API wise, I wonder if it belongs as a shared geoseries/geodataframe method - given the parallels to explode with regards to indexing one might expect the gdf case to return a geodataframe reindexed to correspond to all the coordinates (but on the other hand explode returns a series and not a dataframe).

**Comment by martinfleis** [comment]
This should be ready now.

@m-richards I've moved the exploded index handling into a helper function that is now shared by both explode and get_coordinates. Nice tip! 

> one might expect the gdf case to return a geodataframe reindexed to correspond to all the coordinates

not sure I agree. I would expect it to behave as `.bounds` does. Can limit it to the GeoSeries but not sure about that. Or you'd rather have it behave the same as explode on gdf?

**Comment by m-richards** [comment]
> not sure I agree. I would expect it to behave as `.bounds` does. Can limit it to the GeoSeries but not sure about that. Or you'd rather have it behave the same as explode on gdf?

I actually agree with you - I think the current behaviour makes the most sense in terms of how I might use this method, and the parallel with bounds is a good justification - just thought I'd mention it though so we're happy this is the expected behaviour.

**Review by m-richards — APPROVED** [review]
Looking good to me, a few minor comments.

File: geopandas/base.py [hunk 2]
```diff
@@ -3451,6 +3452,170 @@ def equals(self, other):
...
+
+    Returns
+    -------
+    pandas.Index
+        index or multiindex
+    """
+
+    # the index handling code is resused from GeoSeries.explode
+    if ignore_index:
+        index = range(length)
```
[the remark this answers is not in the document]
Comment (martinfleis) [reply geopandas/base.py]: Good point. Returning None now.

**Comment by martinfleis** [comment]
@jorisvandenbossche can we try to get this in?

**Review by jorisvandenbossche — APPROVED** [review]

**Comment by jorisvandenbossche** [comment]
Thanks!

**Pull request merged**